dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java

```diff
@@ -191,25 +214,36 @@ public RuleValidationResult getDescription( String condition, Program program )
         return ruleEngine.evaluate( condition );
     }
 
-    private RuleEngineContext getRuleEngineContext( Program program, String uid )
+    private RuleEngineContext getRuleEngineContext( Program program, List<ProgramRule> programRules )
     {
         List<ProgramRuleVariable> programRuleVariables = programRuleVariableService
             .getProgramRuleVariable( program );
-        List<ProgramRule> programRules = implementableRuleService
-            .getImplementableRules( program )
-            .stream()
-            .filter( rule -> Objects.isNull( rule.getProgramStage() ) ||
-                Objects.equals( rule.getProgramStage().getUid(), uid ) )
-            .collect( Collectors.toList() );
 
         Map<String, String> constantMap = constantService.getConstantMap().entrySet()
             .stream()
             .collect( Collectors.toMap( Map.Entry::getKey, v -> v.getValue().toString() ) );
 
-        Map<String, List<String>> supplementaryData = organisationUnitGroupService.getAllOrganisationUnitGroups()
-            .stream()
-            .collect( Collectors.toMap( BaseIdentifiableObject::getUid,
-                g -> g.getMembers().stream().map( OrganisationUnit::getUid ).collect( Collectors.toList() ) ) );
+        List<String> orgUnitGroups = new ArrayList<>();
+
+        for ( ProgramRule programRule : programRules )
+        {
+
+            Matcher matcher = PATTERN.matcher( StringUtils.defaultIfBlank( programRule.getCondition(), "" ) );
+
+            while ( matcher.find() )
+            {
+                orgUnitGroups.add( StringUtils.replace( matcher.group( 1 ), "'", "" ) );
+            }
+        }
+
+        Map<String, List<String>> supplementaryData = Maps.newHashMap();
+
+        if ( !orgUnitGroups.isEmpty() )
+        {
+            supplementaryData = orgUnitGroups.stream().collect(
```

Comment:
The rule engine component will only ever need to know the groups for a limited set of org units. Usually there will only be one. We could optimize by only getting the org unit groups of the org units represented as enrollment org unit and event org units in the context data.

Reply:
That is true and it is a better way to solve this.
We could remove the regex dependency and load only the needed data.
I would prefer to develop this in a new PR though, as it can be treated as a separate problem